packages/tre/src/plugins/shared/gateway.ts

```diff
@@ -57,6 +62,8 @@ export class GatewayFactory<Gateway> {
         const unsanitise =
           url.searchParams.get(PARAM_FILE_UNSANITISE) === "true";
         return new FileStorage(root, !unsanitise);
+      } else if (url.protocol === "sqlite:") {
+        return new SqliteStorage(url);
```

Comment:
Multiple `new Database`s should be fine. I'd be hesitant about using `sanitisedNamespace` as the table name, since you parameterise a table name in SQL query, so that would have to be added directly into the string (definitely opening up SQL injection). SQL injection may not be a problem _right now_ since it's all local data and developers can access the database directly anyway, but it's still a bit of a concern. Storing everything in one table with a `namespace` field would seem to be okay?

Reply:
I agree that SQL injection is a concern, which is why I put the namespace as a field. The IDE experience being poor is valid though (although regardless of the solution here, there is only going to be one database file), and should be solved.

My main concern is that we're not able to safely escape the table name, since SQL (whether sqlite or any other engine) doesn't support that. I don't want us to start thinking about doing SQL sanitisation on the namespace ourselves, because that's too easy to accidentally get wrong. Another approach is to run a very strict allowlist regex over the namespace, and use that as the table name, along with a unique suffix (either a GUID or an incrementing number, which could be omitted if the allowlisted name was already unique)? (I'm thinking `replace(/[^a-zA-Z0-9_]+/g, '-'). This namespace-tablename mapping could be stored in a separate table, and wouldn't introduce much overhead, since it would just need to be read/written at database instantiation. The table names wouldn't _exactly_ correspond to namespaces, but they'd be close enough to match by eye in an IDE (`TEST_BUCKET`->`TEST_BUCKET`, `TEST BUCKET`->`TEST-BUCKET`)

Reply:
A hash makes much more sense than a lookup table :joy:

Reply:
Right, that's implemented with a 2 byte hash, which should be good